File: agent/tsdb/src/FeatBit.DataWarehouse/Storage/PathHelper.cs
using System.Text.RegularExpressions;

namespace FeatBit.DataWarehouse.Storage;

/// <summary>
/// Filesystem path utilities shared by <see cref="StorageEngine"/>,
/// <see cref="FlagEvalScanner"/>, and <see cref="MetricEventScanner"/>.
/// </summary>
internal static partial class PathHelper
{
    /// <summary>
    /// Replace any character that is not alphanumeric, hyphen, or underscore with '_'.
    /// Keeps partition directory names safe across Linux, macOS, and Windows.
    /// </summary>
    public static string Sanitize(string input) => InvalidChars().Replace(input, "_");

    [GeneratedRegex(@"[^\w\-]", RegexOptions.CultureInvariant)]
    private static partial Regex InvalidChars();

    // ── Partition path builders ───────────────────────────────────────────────

    public static string FlagEvalPartitionDir(string dataRoot, string envId, string flagKey, string date)
        => Path.Combine(dataRoot, "flag-evals", Sanitize(envId), Sanitize(flagKey), date);

    public static string MetricEventPartitionDir(string dataRoot, string envId, string eventName, string date)
        => Path.Combine(dataRoot, "metric-events", Sanitize(envId), Sanitize(eventName), date);

    /// <summary>
    /// Return all date-partition directories for a flag-eval partition
    /// whose date falls within [startDate, endDate] (inclusive, UTC).
    /// </summary>
    public static IEnumerable<string> FlagEvalDateDirs(
        string dataRoot, string envId, string flagKey,
        DateOnly startDate, DateOnly endDate)
        => DateDirs(
            Path.Combine(dataRoot, "flag-evals", Sanitize(envId), Sanitize(flagKey)),
            startDate, endDate);

    /// <summary>
    /// Return all date-partition directories for a metric-event partition
    /// whose date falls within [startDate, endDate] (inclusive, UTC).
    /// </summary>
    public static IEnumerable<string> MetricEventDateDirs(
        string dataRoot, string envId, string eventName,
        DateOnly startDate, DateOnly endDate)
        => DateDirs(
            Path.Combine(dataRoot, "metric-events", Sanitize(envId), Sanitize(eventName)),
            startDate, endDate);

    private static IEnumerable<string> DateDirs(string root, DateOnly startDate, DateOnly endDate)
    {
        if (!Directory.Exists(root)) yield break;

        foreach (var dir in Directory.EnumerateDirectories(root).Order())
        {
            var name = Path.GetFileName(dir);
            if (DateOnly.TryParseExact(name, "yyyy-MM-dd", out var date)
                && date >= startDate && date <= endDate)
            {
                yield return dir;
            }
        }
    }
}

File: agent/tsdb-cloudflare/src/storage/path-helper.ts
/**
 * R2 key construction and date-range helpers.
 * Mirrors .NET PathHelper — translates filesystem paths to R2 object key prefixes.
 */

/** Replace any char that is not word char or hyphen with '_'. */
export function sanitize(input: string): string {
  return input.replace(/[^\w-]/g, "_");
}

// ── Partition key (prefix) builders ───────────────────────────────────────────

export function flagEvalPrefix(envId: string, flagKey: string, date: string): string {
  return `flag-evals/${sanitize(envId)}/${sanitize(flagKey)}/${date}/`;
}

export function metricEventPrefix(envId: string, eventName: string, date: string): string {
  return `metric-events/${sanitize(envId)}/${sanitize(eventName)}/${date}/`;
}

// ── Date range enumeration ────────────────────────────────────────────────────

/** Generate all yyyy-MM-dd strings in [startDate, endDate] inclusive. */
export function dateRange(startDate: string, endDate: string): string[] {
  const dates: string[] = [];
  const start = new Date(startDate + "T00:00:00Z");
  const end = new Date(endDate + "T00:00:00Z");

  for (let d = new Date(start); d <= end; d.setUTCDate(d.getUTCDate() + 1)) {
    dates.push(d.toISOString().slice(0, 10));
  }
  return dates;
}

/** List all R2 prefixes for flag-eval segments across a date range. */
export function flagEvalPrefixes(
  envId: string, flagKey: string, startDate: string, endDate: string,
): string[] {
  return dateRange(startDate, endDate).map(d => flagEvalPrefix(envId, flagKey, d));
}

/** List all R2 prefixes for metric-event segments across a date range. */
export function metricEventPrefixes(
  envId: string, eventName: string, startDate: string, endDate: string,
): string[] {
  return dateRange(startDate, endDate).map(d => metricEventPrefix(envId, eventName, d));
}

// ── Rollup paths ──────────────────────────────────────────────────────────────

export function flagEvalRollupKey(envId: string, flagKey: string, date: string): string {
  return `rollups/flag-evals/${sanitize(envId)}/${sanitize(flagKey)}/${date}.json`;
}

export function metricEventRollupKey(envId: string, eventName: string, date: string): string {
  return `rollups/metric-events/${sanitize(envId)}/${sanitize(eventName)}/${date}.json`;
}

export function flagEvalRollupPrefix(envId: string, flagKey: string): string {
  return `rollups/flag-evals/${sanitize(envId)}/${sanitize(flagKey)}/`;
}

export function metricEventRollupPrefix(envId: string, eventName: string): string {
  return `rollups/metric-events/${sanitize(envId)}/${sanitize(eventName)}/`;
}
